buildbot/sycl-bisect.bash

```diff
@@ -0,0 +1,195 @@
+      exit 1
+      ;;
+  esac
+done
+
+# Validate options.
+if [[ "$1" == "" ]] || [[ "$2" == "" ]]; then
+  echo "A bad rev and at least one good rev must be passed" >&2
+  exit 1
+fi
+if [[ "$TEST" == "" ]] && [[ "$COMMAND" == "" ]]; then
+  echo "--test <test> or --command <command> must be used to specify a test" >&2
+  exit 1
+fi
+if [[ "$TEST" != "" ]] && [[ "$COMMAND" != "" ]]; then
+  echo "Only one of --test <test> or --command <command> is allowed" >&2
+  exit 1
+fi
+if [[ "$COMMAND_ALLOW_BISECT_CODES" != "" ]] && [[ "$COMMAND" == "" ]]; then
+  echo "--command-allow-bisect-codes is only allowed with --command <command>" >&2
+  exit 1
+fi
+
+# Save the current working directory before switching to the repository's
+# top-level directory. Tests will be run relative to the original working
+# directory.
+export TEST_WD="$PWD"
+TOPLEVEL="$(git rev-parse --show-toplevel)"
+if [[ "$TOPLEVEL" == "" ]]; then
+  exit 1
+fi
+cd "$TOPLEVEL"
+
+# A commit that should be on all sycl-based branches this script works with.
+# This is the commit that introduces the buildbot scripts, which this script
+# uses for building.
+export SYCL_ANCESTOR=0818f3e88585
+
+# This commit should be in the repository.
+if ! git rev-parse $SYCL_ANCESTOR &>/dev/null; then
+  echo "Current repository is not sycl-based" >&2
+  exit 1
+fi
+
+# Along with all the bad/good revs.
+for REV in "$@"; do
+  if ! git rev-parse --verify "$REV" &>/dev/null; then
+    echo "'$REV' is not a valid revision" >&2
+    exit 1
+  fi
+done
+
+# Make sure the build directory is set up; if not, tell the user to run
+# configure.py.
+if [[ ! -f build/CMakeCache.txt ]]; then
+  echo "The build directory doesn't seem to be configured yet." >&2
+  echo "Please run buildbot/configure.py" >&2
+  exit 1
+fi
+
+# If this is a LIT test, make sure FileCheck and other testing utilities are
+# built.
+if [[ ! -f build/bin/FileCheck ]]; then
+  echo "FileCheck not found; building test-depends"
+  python buildbot/compile.py -t test-depends
+fi
+
+# Determines if the passed commit is sycl-based or not.
+function is_sycl_based_commit {
+  git merge-base --is-ancestor $SYCL_ANCESTOR "$1" 2>/dev/null
+}
+
+# Checks if this commit/branch is a candidate for SYCL_DESCENDANT. For that to
+# be the case, it needs to be sycl-based and a descendant of the bad and good
+# commits.
+readonly -a REQUIRED_ANCESTORS=($SYCL_ANCESTOR "$@")
+function check_sycl_descendant {
+  local ancestor
+  for ancestor in "${REQUIRED_ANCESTORS[@]}"; do
+    git merge-base --is-ancestor "$ancestor" "$1" &>/dev/null || return 1
+  done
+  export SYCL_DESCENDANT="$1"
+  return 0
+}
+
+# If SYCL_DESCENDANT isn't specified, poke around until a suitable commit/branch
+# is found.
+function find_sycl_descendant {
+  [[ "$SYCL_DESCENDANT" != "" ]] && return
+
+  # Try the bad rev first.
+  check_sycl_descendant "${REQUIRED_ANCESTORS[1]}" && return
+
+  # Try "sycl" and "sycl-web", both locally and in all the remotes.
```

Comment:
nit: should we limit that to official remotes only? If, for some reason, somebody uses ordinary development workspace there might be extra remotes that wouldn't make sense (e.g. my github remote has *very* old sycl/sycl-web).

Reply:
This is just looking for _any_ commit that is descended from REQUIRED_ANCESTORS, so extra remotes shouldn't cause any issues besides slowing down this check. This is part of an elaborate workaround to try to approximate a version of git-merge-base that finds common descendants instead of ancestors, which is usually not a well-defined question in git because you can have commits merged differently in different branches. In our usage that's very unlikely, though, so any branch we can find that's descended from the required commits is good enough for figuring out where to merge. We still probably don't want to test _every_ remote branch because there can be a lot of them and it would take forever to find out if none of them have all of the commits, but just checking the "sycl" and "sycl-web" branches of each remote seems like a good compromise to still be able to catch all of the most likely remote branches.